kinova_driver/src/kinova_arm.cpp

```diff
@@ -0,0 +1,885 @@
+    l_joint_torque_[3] = joint_torque->joint4;
+    l_joint_torque_[4] = joint_torque->joint5;
+    l_joint_torque_[5] = joint_torque->joint6;
+    l_joint_torque_[6] = joint_torque->joint7;
+
+    kinova_comm_.setJointTorques(l_joint_torque_);
+  }
+}
+
+void KinovaArm::forceSubscriberCallback(const kinova_msgs::CartesianForceConstPtr& force)
+{
+  if (!kinova_comm_.isStopped())
+  {
+    l_force_cmd_[0] = force->force_x;
+    l_force_cmd_[1] = force->force_y;
+    l_force_cmd_[2] = force->force_z;
+    l_force_cmd_[3] = force->torque_x;
+    l_force_cmd_[4] = force->torque_y;
+    l_force_cmd_[5] = force->torque_z;
+
+    kinova_comm_.sendCartesianForceCommand(l_force_cmd_);
+  }
+}
+
+void KinovaArm::OvisJointVelocityCallback(const ovis_msgs::OvisJointVelocity::ConstPtr& msg)
+{
+  this->trajectory_point.Position.Actuators.Actuator1 = 0;
+  this->trajectory_point.Position.Actuators.Actuator2 = 0;
+  this->trajectory_point.Position.Actuators.Actuator3 = 0;
+  this->trajectory_point.Position.Actuators.Actuator4 = 0;
+  this->trajectory_point.Position.Actuators.Actuator5 = 0;
+  this->trajectory_point.Position.Actuators.Actuator6 = 0;
+  switch (msg->joint_index)
+  {
+    case 0:
+      trajectory_point.Position.Actuators.Actuator1 = msg->joint_velocity * number_of_degree_per_sec;
+      break;
+    case 1:
+      trajectory_point.Position.Actuators.Actuator2 = INVERSE * msg->joint_velocity * number_of_degree_per_sec;
+      break;
+    case 2:
+      trajectory_point.Position.Actuators.Actuator3 = msg->joint_velocity * number_of_degree_per_sec;
+      break;
+    case 3:
+      trajectory_point.Position.Actuators.Actuator4 = msg->joint_velocity * number_of_degree_per_sec;
+      break;
+    case 4:
+      trajectory_point.Position.Actuators.Actuator5 = msg->joint_velocity * number_of_degree_per_sec;
+      break;
+    case 5:
+      trajectory_point.Position.Actuators.Actuator6 = msg->joint_velocity * number_of_degree_per_sec;
+      break;
+  }
+  kinova_comm_.SendBasicTrajectoryVelocity(trajectory_point);
+}
+
+void KinovaArm::OvisJointAngleCallback(const ovis_msgs::OvisJointAngle::ConstPtr& msg)
+{
+  this->trajectory_point.Position.Actuators.Actuator1 = msg->joint_angles[0];
+  this->trajectory_point.Position.Actuators.Actuator2 = msg->joint_angles[1];
+  this->trajectory_point.Position.Actuators.Actuator3 = msg->joint_angles[2];
+  this->trajectory_point.Position.Actuators.Actuator4 = msg->joint_angles[3];
+  this->trajectory_point.Position.Actuators.Actuator5 = msg->joint_angles[4];
+  this->trajectory_point.Position.Actuators.Actuator6 = msg->joint_angles[5];
+
+  kinova_comm_.SendBasicTrajectoryPosition(trajectory_point);
+}
+
+bool KinovaArm::HomePositionSrvCallback(std_srvs::Trigger::Request& req, std_srvs::Trigger::Response& res)
+{
+  kinova_comm_.SendBasicTrajectoryPosition(home_trajectory_point);
+  res.message = "successfully send home position";  // TODO add a check for the home position
```

Comment:
What does that TODO mean? Maybe add task to Teams?

Reply:
I think what I meant was that we ought to add a check to confirm that the position of the joint is really at the position we told them to go with the HomePositionSrvCallback, since I had weird behaviour with the callback when testing. I'll remove the comment for now, but I'll test it before merging and if a check need to be added I'll add one.

Reply:
For the message, change "send" to "sent".